src/vmm/src/vmm_config/drive.rs

```diff
@@ -70,109 +81,114 @@ pub struct BlockDeviceConfig {
-    pub fn new() -> BlockDeviceConfigs {
-        BlockDeviceConfigs {
-            config_list: VecDeque::<BlockDeviceConfig>::new(),
+impl BlockDevices {
+    /// Constructor for BlockDevices. It initializes an empty LinkedList.
+    pub fn new() -> BlockDevices {
+        BlockDevices {
+            list: VecDeque::<Arc<Mutex<Block>>>::new(),
             has_root_block: false,
         }
     }
 
     /// Gets the index of the device with the specified `drive_id` if it exists in the list.
     pub fn get_index_of_drive_id(&self, drive_id: &str) -> Option<usize> {
-        self.config_list
-            .iter()
-            .position(|cfg| cfg.drive_id.eq(drive_id))
-    }
-
-    fn get_index_of_drive_path(&self, drive_path: &PathBuf) -> Option<usize> {
-        self.config_list
+        self.list
             .iter()
-            .position(|cfg| cfg.path_on_host.eq(drive_path))
+            .position(|b| b.lock().unwrap().id().eq(drive_id))
     }
 
-    /// Inserts `block_device_config` in the block device configuration list.
-    /// If an entry with the same id already exists, it will attempt to update
-    /// the existing entry.
+    /// Inserts a `Block` in the block devices list using the specified configuration.
+    /// If a block with the same id already exists, it will overwrite it.
     /// Inserting a secondary root block device will fail.
-    pub fn insert(&mut self, block_device_config: BlockDeviceConfig) -> Result<()> {
-        // If the id of the drive already exists in the list, the operation is update.
-        match self.get_index_of_drive_id(&block_device_config.drive_id) {
-            Some(index) => self.update(index, block_device_config),
-            None => self.create(block_device_config),
-        }
-    }
-
-    fn create(&mut self, block_device_config: BlockDeviceConfig) -> Result<()> {
-        // check if the path exists
-        if !block_device_config.path_on_host.exists() {
-            return Err(DriveError::InvalidBlockDevicePath);
-        }
-
-        if self
-            .get_index_of_drive_path(&block_device_config.path_on_host)
-            .is_some()
-        {
-            return Err(DriveError::BlockDevicePathAlreadyExists);
-        }
-
-        // check whether the Device Config belongs to a root device
-        // we need to satisfy the condition by which a VMM can only have on root device
-        if block_device_config.is_root_device {
-            if self.has_root_block {
-                return Err(DriveError::RootBlockDeviceAlreadyAdded);
-            } else {
-                self.has_root_block = true;
-                // Root Device should be the first in the list whether or not PARTUUID is specified
-                // in order to avoid bugs in case of switching from partuuid boot scenarios to
-                // /dev/vda boot type.
-                self.config_list.push_front(block_device_config);
+    pub fn insert(&mut self, config: BlockDeviceConfig) -> Result<()> {
```

Comment:
Would it make sense to use this path when restoring devices from state?. One option of doing so would be to make config an Enum with 2 variants : BlockDeviceConfig, BlockState. I am not sure if this would work out fine given the fact that we also need other stuff like guest memory.